File: C++/538. Convert BST to Greater Tree.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 // class Solution {
 // public:
 //     int sum = 0;
 //     void traversal(TreeNode* root){
 //         if (!root) return;
 //         traversal(root->right);
 //         root->val += sum;
 //         sum = root->val;
 //         traversal(root->left);
 //     }
 //     TreeNode* convertBST(TreeNode* root) {
 //         sum = 0;
 //         traversal(root);
 //         return root;
 //     }
 // };

class Solution {
public:
    int sum = 0;
    TreeNode* convertBST(TreeNode* root) {
        if (!root) return root;
        stack<TreeNode*> st;
        st.push(root);
        sum = 0;
        while (!st.empty()) {
            TreeNode* node = st.top(); st.pop();
            if (node) {
                if (node->left) st.push(node->left);
                st.push(node);
                st.push(NULL);
                if (node->right) st.push(node->right);
            }
            else {
                TreeNode* node = st.top(); st.pop();
                node->val += sum;
                sum = node->val;
            }
        }
        return root;
    }
};

File: Python/538. Convert BST to Greater Tree.py
'''
1.递归
2.迭代
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
#         sum_ = 0
#         def traversal(root):
#             nonlocal sum_
#             if not root:
#                 return
#             traversal(root.right)
#             root.val += sum_
#             sum_ = root.val
#             traversal(root.left)
#         traversal(root)
#         return root

class Solution:
    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return root
        st = [root]
        sum_ = 0
        while st:
            node = st.pop(-1)
            if node:
                if node.left: st.append(node.left)
                st.append(node)
                st.append(None)
                if node.right: st.append(node.right)
            else:
                node = st.pop(-1)
                node.val += sum_
                sum_ = node.val
        return root